documentation/routing/tutorial-router-templates.asciidoc

```diff
@@ -0,0 +1,382 @@
+
+* Resolved examples:
+** `api/com/vaadin/flow`, where parameter `path` has value `"com/vaadin/flow"`.
+** `api/com/flow`, where parameter `path` has value `"com/flow"`
+** `api/flow/vaadin`, where parameter `path` has value `"flow/vaadin"`
+* Unresolved example:
+** `api/com/vaadin/framework`.
+
+[source,java]
+----
+@Route("api/:path*(com|vaadin|flow)")
+public class ApiViewer extends Div implements BeforeEnterObserver {
+}
+----
+
+[NOTE]
+Regarding optional parameters which are greedily matched from left to right, giving the template `path/to/:param1?([0-9]\*)/:param2?([a-z]*)` following urls match it:
+
+* `path/to` with no parameter,
+* `path/to/123`, where `param1` = `123`,
+* `path/to/123/qwe`, where `param1` = `123` and `param2` = `qwe`,
+
+while `path/to/qwe/123` do not match the template.
+
+== Route template priority
+
+For an application with a complex structure, the list of route templates may bring some overlapping in the definition of parameters for each route.
+
+The Router engine will deny by default any attempt to register the same route for more than one view.
+Also a route containing optional parameters is in conflict with the same route without the parameters and the last to be register will fail.
+The failure consists in a `InvalidRouteConfigurationException` being thrown during route registration leading to the termination of the application.
+
+*Example*: Following configuration will fail since both are resolving `items/show` and this is obvious at configuration time.
+
+[source,java]
+----
+@Route("items/show")
+public static class ShowAllView extends Div {
+}
+
+// This route will fail when registered and application is terminated.
+@Route("items/show/:filter?")
+public static class SearchView extends Div {
+}
+----
+
+[NOTE]
+One way to fix this is to make `filter` parameter mandatory, by removing the optional modifier.
+The resulted route will look like `@Route("items/show/:filter")`.
+The other possibility is to remove `ShowAllView` class and show all items using `SearchView` when the `filter` parameter is missing.
+
+However, since identifying all possible ambiguities between route templates is computationally intractable, a priority mechanism has to be used when the url is resolved, instead of failing the application when a conflicting route is registered.
+Thus, depending on the parameter modifier and the order the routes are registered, one route has priority over the others.
+This is applicable for any defined route, on the same navigation view or another view, and using both `@Route` or `@RouteAlias`.
+
+This is the order is which segments in the registered route templates are trying to resolve the url:
+
+1. Static segment.
+2. Mandatory parameter.
+3. Optional parameter.
+4. Next segments following the optional parameter.
+5. Wildcard parameter.
+
```

Comment:
What would be an order for regexp parameters?

Reply:
Regexp is an additional match criterion for parameters, it does not as such influence the priority order.